kubernetes/cache.go

```diff
@@ -454,3 +670,118 @@ func (c *controllerImpl) GetServices(namespace string) ([]core_v1.Service, error
 	}
 	return []core_v1.Service{}, nil
 }
+
+func (c *controllerImpl) GetProjects() ([]osproject_v1.Project, error) {
+	if err := c.checkStateAndRetry(); err != nil {
+		return []osproject_v1.Project{}, err
+	}
+
+	projects := c.projectInformer.GetIndexer().List()
+	if len(projects) > 0 {
+		if _, ok := projects[0].(*osproject_v1.Project); ok {
+			osProjects := make([]osproject_v1.Project, len(projects))
+			for i, project := range projects {
+				osProjects[i] = *(project.(*osproject_v1.Project))
+			}
+			return osProjects, nil
+		}
+	}
+
+	return []osproject_v1.Project{}, nil
+}
+
+func (c *controllerImpl) GetNamespaces() ([]core_v1.Namespace, error) {
+	if err := c.checkStateAndRetry(); err != nil {
+		return []core_v1.Namespace{}, err
+	}
+
+	namespaces := c.projectInformer.GetIndexer().List()
+	if len(namespaces) > 0 {
+		if _, ok := namespaces[0].(*core_v1.Namespace); ok {
+			coreNamespaces := make([]core_v1.Namespace, len(namespaces))
+			for i, ns := range namespaces {
+				coreNamespaces[i] = *(ns.(*core_v1.Namespace))
+			}
+			return coreNamespaces, nil
+		}
+	}
+
+	return []core_v1.Namespace{}, nil
+
+}
+
+func createProjectsInformer(client *osproject_v1_client.ProjectV1Client, refreshDuration time.Duration) cache.SharedIndexInformer {
+	return cache.NewSharedIndexInformer(cache.NewListWatchFromClient(client.RESTClient(), "projects", meta_v1.NamespaceAll, fields.Everything()),
+		&osproject_v1.Project{},
+		refreshDuration,
+		cache.Indexers{},
+	)
+}
+
+func createNamespaceInformer(clientset kubernetes.Interface, refreshDuration time.Duration) cache.SharedIndexInformer {
+	sharedInformer := informers.NewSharedInformerFactory(clientset, refreshDuration)
+	return sharedInformer.Core().V1().Namespaces().Informer()
+}
+
+func (c *controllerImpl) GetVirtualServices(namespace string) (*GenericIstioObjectList, error) {
+	return c.getGenericIstioObjectList(namespace, virtualServiceType)
+}
+
+func (c *controllerImpl) getGenericIstioObjectList(namespace, istioType string) (*GenericIstioObjectList, error) {
+	// Do the parsing from GenericIstioObjectList to the correct one in istio_details_service.go so we don't replicate that code..
+	if err := c.checkStateAndRetry(); err != nil {
+		return nil, err
+	}
+	if _, found := c.istioControllers[namespace]; !found {
+		log.Debugf("No %s informers for %s found", istioType, namespace) // Debugf
+		return nil, fmt.Errorf("Istio caching is not enabled for namespace %s", namespace)
+	}
+
+	vss := c.istioControllers[namespace][istioType].GetIndexer().List()
+	log.Infof("Len of vss: %d for %s\n", len(vss), namespace)
```

Comment:
Minor. as this is generic, the vss perhaps can be changed for the type it is fetching.